// @extrimian/did-core/dist/models/purposes.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyAgreementPurpose = exports.CapabilityInvocationPurpose = exports.AssertionMethodPurpose = exports.AuthenticationPurpose = exports.Purpose = void 0;
var Purpose = /** @class */ (function () {
    function Purpose() {
    }
    return Purpose;
}());
exports.Purpose = Purpose;
var AuthenticationPurpose = /** @class */ (function (_super) {
    __extends(AuthenticationPurpose, _super);
    function AuthenticationPurpose(init) {
        var _this = _super.call(this) || this;
        _this.name = "authentication";
        Object.assign(_this, init);
        return _this;
    }
    return AuthenticationPurpose;
}(Purpose));
exports.AuthenticationPurpose = AuthenticationPurpose;
var AssertionMethodPurpose = /** @class */ (function () {
    function AssertionMethodPurpose() {
        this.name = "assertionMethod";
    }
    return AssertionMethodPurpose;
}());
exports.AssertionMethodPurpose = AssertionMethodPurpose;
var CapabilityInvocationPurpose = /** @class */ (function () {
    function CapabilityInvocationPurpose() {
        this.name = "capabilityInvocation";
    }
    return CapabilityInvocationPurpose;
}());
exports.CapabilityInvocationPurpose = CapabilityInvocationPurpose;
var KeyAgreementPurpose = /** @class */ (function () {
    function KeyAgreementPurpose() {
        this.name = "keyAgreement";
    }
    return KeyAgreementPurpose;
}());
exports.KeyAgreementPurpose = KeyAgreementPurpose;
