// @extrimian/waci/dist/handlers/issuance/step-5-request-credential.handler.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestCredentialHandler = void 0;
var lodash_1 = require("lodash");
var register_handler_decorator_1 = require("../decorators/register-handler.decorator");
var types_1 = require("../../types");
var utils_1 = require("../../utils");
var callbacks_1 = require("../../callbacks");
var problem_report_1 = require("../../types/problem-report");
var RequestCredentialHandler = /** @class */ (function () {
    function RequestCredentialHandler() {
    }
    RequestCredentialHandler.prototype.handle = function (messageThread) {
        var _a, e_1, _b, _c;
        var _d;
        return __awaiter(this, void 0, void 0, function () {
            var messageToProcess, holderDID, issuerDID, problemReport, offerCredentialMessage, manifests, applicationsToCheck, verificationResultCallback, vcs, _e, applicationsToCheck_1, applicationsToCheck_1_1, applicationToCheck, verify, result, _i, _f, vc, challengeToCheck, presentation, verifyPresentationResult, e_1_1, fulfillments;
            var _g, _h;
            var _this = this;
            return __generator(this, function (_j) {
                switch (_j.label) {
                    case 0:
                        messageToProcess = messageThread[messageThread.length - 1];
                        holderDID = messageToProcess.from;
                        issuerDID = messageToProcess.to[0];
                        problemReport = new problem_report_1.ProblemReportMessage();
                        offerCredentialMessage = messageThread.find(function (message) { return message.type === types_1.WACIMessageType.OfferCredential; });
                        manifests = offerCredentialMessage.attachments.filter(function (attachment) { var _a, _b; return !(0, lodash_1.isNil)((_b = (_a = attachment === null || attachment === void 0 ? void 0 : attachment.data) === null || _a === void 0 ? void 0 : _a.json) === null || _b === void 0 ? void 0 : _b.credential_manifest); });
                        applicationsToCheck = manifests
                            .filter(function (manifest) {
                            var _a, _b, _c;
                            return !(0, lodash_1.isNil)((_c = (_b = (_a = manifest === null || manifest === void 0 ? void 0 : manifest.data) === null || _a === void 0 ? void 0 : _a.json) === null || _b === void 0 ? void 0 : _b.credential_manifest) === null || _c === void 0 ? void 0 : _c.presentation_definition);
                        })
                            .map(function (manifest) { return ({
                            presentationDefinition: manifest.data.json.credential_manifest.presentation_definition,
                            application: messageToProcess.attachments.find(function (attachment) {
                                return attachment.data.json.credential_application.manifest_id ===
                                    manifest.data.json.credential_manifest.id;
                            }),
                        }); });
                        verificationResultCallback = (_d = callbacks_1.callbacks[types_1.Actor.Issuer]) === null || _d === void 0 ? void 0 : _d.credentialVerificationResult;
                        vcs = [];
                        if (!applicationsToCheck.every(function (applicationToCheck) {
                            return !(0, lodash_1.isNil)(applicationToCheck.application);
                        })) return [3 /*break*/, 18];
                        _j.label = 1;
                    case 1:
                        _j.trys.push([1, 11, 12, 17]);
                        _e = true, applicationsToCheck_1 = __asyncValues(applicationsToCheck);
                        _j.label = 2;
                    case 2: return [4 /*yield*/, applicationsToCheck_1.next()];
                    case 3:
                        if (!(applicationsToCheck_1_1 = _j.sent(), _a = applicationsToCheck_1_1.done, !_a)) return [3 /*break*/, 10];
                        _c = applicationsToCheck_1_1.value;
                        _e = false;
                        _j.label = 4;
                    case 4:
                        _j.trys.push([4, , 8, 9]);
                        applicationToCheck = _c;
                        return [4 /*yield*/, (0, utils_1.verifyPresentation)(applicationToCheck.presentationDefinition, applicationToCheck.application, callbacks_1.callbacks[types_1.Actor.Issuer].verifyCredential)];
                    case 5:
                        verify = _j.sent();
                        result = verify.result;
                        for (_i = 0, _f = verify.vcs; _i < _f.length; _i++) {
                            vc = _f[_i];
                            vcs.push(vc);
                        }
                        if (!result) {
                            if (verificationResultCallback) {
                                verificationResultCallback({
                                    result: verify.result,
                                    error: verify.erorrs,
                                    thid: messageToProcess.thid,
                                    vcs: verify.vcs,
                                    message: messageToProcess,
                                });
                            }
                            return [2 /*return*/, {
                                    responseType: 1 /* WACIMessageResponseType.ReplyThread */,
                                    message: {
                                        type: types_1.WACIMessageType.ProblemReport,
                                        id: (0, utils_1.createUUID)(),
                                        thid: messageToProcess.id,
                                        from: issuerDID,
                                        to: [holderDID],
                                        body: problemReport.presentProofMessage(verify.error.name, verify.error.description),
                                    },
                                }];
                        }
                        if (!applicationsToCheck.length) return [3 /*break*/, 7];
                        challengeToCheck = (0, utils_1.extractExpectedChallenge)(offerCredentialMessage);
                        presentation = messageToProcess.attachments[0].data.json;
                        return [4 /*yield*/, callbacks_1.callbacks[types_1.Actor.Issuer].verifyPresentation({ presentation: presentation, challenge: challengeToCheck })];
                    case 6:
                        verifyPresentationResult = _j.sent();
                        if (!verifyPresentationResult.result) {
                            return [2 /*return*/, {
                                    responseType: 1 /* WACIMessageResponseType.ReplyThread */,
                                    message: {
                                        type: types_1.WACIMessageType.ProblemReport,
                                        id: (0, utils_1.createUUID)(),
                                        thid: messageToProcess.id,
                                        from: issuerDID,
                                        to: [holderDID],
                                        body: problemReport.presentProofMessage(verifyPresentationResult.error.name, verifyPresentationResult.error.description),
                                    },
                                }];
                        }
                        _j.label = 7;
                    case 7: return [3 /*break*/, 9];
                    case 8:
                        _e = true;
                        return [7 /*endfinally*/];
                    case 9: return [3 /*break*/, 2];
                    case 10: return [3 /*break*/, 17];
                    case 11:
                        e_1_1 = _j.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 17];
                    case 12:
                        _j.trys.push([12, , 15, 16]);
                        if (!(!_e && !_a && (_b = applicationsToCheck_1.return))) return [3 /*break*/, 14];
                        return [4 /*yield*/, _b.call(applicationsToCheck_1)];
                    case 13:
                        _j.sent();
                        _j.label = 14;
                    case 14: return [3 /*break*/, 16];
                    case 15:
                        if (e_1) throw e_1.error;
                        return [7 /*endfinally*/];
                    case 16: return [7 /*endfinally*/];
                    case 17:
                        if (verificationResultCallback) {
                            verificationResultCallback({
                                result: true,
                                error: null,
                                thid: messageToProcess.thid,
                                vcs: vcs,
                                message: messageToProcess,
                            });
                        }
                        _j.label = 18;
                    case 18:
                        fulfillments = offerCredentialMessage.attachments.filter(function (attachment) { var _a, _b; return !(0, lodash_1.isNil)((_b = (_a = attachment === null || attachment === void 0 ? void 0 : attachment.data) === null || _a === void 0 ? void 0 : _a.json) === null || _b === void 0 ? void 0 : _b.credential_fulfillment); });
                        _g = {
                            responseType: 1 /* WACIMessageResponseType.ReplyThread */
                        };
                        _h = {
                            type: types_1.WACIMessageType.IssueCredential,
                            id: (0, utils_1.createUUID)(),
                            thid: messageToProcess.thid,
                            from: issuerDID,
                            to: [holderDID],
                            body: {}
                        };
                        return [4 /*yield*/, Promise.all(fulfillments.map(function (_a) {
                                var json = _a.data.json;
                                return __awaiter(_this, void 0, void 0, function () {
                                    var _b;
                                    var _c, _d, _e;
                                    return __generator(this, function (_f) {
                                        switch (_f.label) {
                                            case 0:
                                                _c = {
                                                    id: (0, utils_1.createUUID)(),
                                                    media_type: 'application/json',
                                                    format: 'dif/credential-manifest/fulfillment@v1.0'
                                                };
                                                _d = {};
                                                _b = [__assign({}, json)];
                                                _e = {};
                                                return [4 /*yield*/, Promise.all(json.verifiableCredential.map(function (vc) {
                                                        return callbacks_1.callbacks[types_1.Actor.Issuer].signCredential({ vc: vc, message: messageToProcess });
                                                    }))];
                                            case 1: return [2 /*return*/, (_c.data = (_d.json = __assign.apply(void 0, _b.concat([(_e.verifiableCredential = _f.sent(), _e)])),
                                                    _d),
                                                    _c)];
                                        }
                                    });
                                });
                            }))];
                    case 19: return [2 /*return*/, (_g.message = (_h.attachments = _j.sent(),
                            _h),
                            _g)];
                }
            });
        });
    };
    RequestCredentialHandler = __decorate([
        (0, register_handler_decorator_1.RegisterHandler)(types_1.Actor.Issuer, types_1.WACIMessageType.RequestCredential)
    ], RequestCredentialHandler);
    return RequestCredentialHandler;
}());
exports.RequestCredentialHandler = RequestCredentialHandler;
