// @extrimian/waci/dist/handlers/issuance/step-4-offer-credential.handler.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OfferCredentialHandler = void 0;
var register_handler_decorator_1 = require("../decorators/register-handler.decorator");
var types_1 = require("../../types");
var utils_1 = require("../../utils");
var callbacks_1 = require("../../callbacks");
var lodash_1 = require("lodash");
var OfferCredentialHandler = /** @class */ (function () {
    function OfferCredentialHandler() {
    }
    OfferCredentialHandler.prototype.handle = function (messageThread) {
        return __awaiter(this, void 0, void 0, function () {
            var message, holderDID, issuerDID, credentialApplicationParams, _a, credentialsToPresent, presentationProofTypes, credentialApplication;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        message = messageThread[messageThread.length - 1];
                        holderDID = message.to[0];
                        issuerDID = message.from;
                        credentialApplicationParams = {
                            manifest: message.attachments.find(function (attachment) { var _a, _b; return !(0, lodash_1.isNil)((_b = (_a = attachment === null || attachment === void 0 ? void 0 : attachment.data) === null || _a === void 0 ? void 0 : _a.json) === null || _b === void 0 ? void 0 : _b.credential_manifest); }),
                            fulfillment: message.attachments.find(function (attachment) { var _a, _b; return !(0, lodash_1.isNil)((_b = (_a = attachment === null || attachment === void 0 ? void 0 : attachment.data) === null || _a === void 0 ? void 0 : _a.json) === null || _b === void 0 ? void 0 : _b.credential_fulfillment); }),
                            message: message
                        };
                        if (!credentialApplicationParams.manifest)
                            throw new Error('Malformed offer credential message');
                        return [4 /*yield*/, callbacks_1.callbacks[types_1.Actor.Holder].getCredentialApplication(credentialApplicationParams)];
                    case 1:
                        _a = _b.sent(), credentialsToPresent = _a.credentialsToPresent, presentationProofTypes = _a.presentationProofTypes;
                        return [4 /*yield*/, this.createMessage(credentialsToPresent, presentationProofTypes, credentialApplicationParams.manifest, message)];
                    case 2:
                        credentialApplication = _b.sent();
                        return [2 /*return*/, {
                                responseType: 1 /* WACIMessageResponseType.ReplyThread */,
                                message: {
                                    type: types_1.WACIMessageType.RequestCredential,
                                    id: (0, utils_1.createUUID)(),
                                    thid: message.thid,
                                    from: holderDID,
                                    to: [issuerDID],
                                    body: {},
                                    attachments: credentialApplication ? [credentialApplication] : [],
                                },
                            }];
                }
            });
        });
    };
    OfferCredentialHandler.prototype.createMessage = function (verifiableCredential, proofTypes, manifest, message) {
        var _a, _b, _c, _d, _e;
        return __awaiter(this, void 0, void 0, function () {
            var credential_manifest, manifest_id, presentation_definition, definition_id, input_descriptors, challenge, descriptor_map, messageData, signedData;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        credential_manifest = (_b = (_a = manifest === null || manifest === void 0 ? void 0 : manifest.data) === null || _a === void 0 ? void 0 : _a.json) === null || _b === void 0 ? void 0 : _b.credential_manifest;
                        manifest_id = credential_manifest === null || credential_manifest === void 0 ? void 0 : credential_manifest.id;
                        presentation_definition = credential_manifest === null || credential_manifest === void 0 ? void 0 : credential_manifest.presentation_definition;
                        definition_id = presentation_definition === null || presentation_definition === void 0 ? void 0 : presentation_definition.id;
                        input_descriptors = presentation_definition === null || presentation_definition === void 0 ? void 0 : presentation_definition.input_descriptors;
                        challenge = (_e = (_d = (_c = manifest === null || manifest === void 0 ? void 0 : manifest.data) === null || _c === void 0 ? void 0 : _c.json) === null || _d === void 0 ? void 0 : _d.options) === null || _e === void 0 ? void 0 : _e.challenge;
                        if (!challenge)
                            throw new Error('No challenge defined');
                        if (!(input_descriptors === null || input_descriptors === void 0 ? void 0 : input_descriptors.length))
                            return [2 /*return*/];
                        descriptor_map = input_descriptors === null || input_descriptors === void 0 ? void 0 : input_descriptors.map(function (descriptor, index) { return ({
                            id: descriptor.id,
                            format: 'ldp_vp',
                            path: "$.verifiableCredential[".concat(index, "]"),
                        }); });
                        if (!(proofTypes === null || proofTypes === void 0 ? void 0 : proofTypes.length))
                            throw new Error('Proof types need to be defined');
                        if (!manifest_id)
                            throw new Error('Malformed credential manifest');
                        messageData = {
                            '@context': [
                                'https://extrimian.blob.core.windows.net/rskec/securityv1.jsonld',
                                'https://extrimian.blob.core.windows.net/rskec/credentialsv1.jsonld',
                                'https://extrimian.blob.core.windows.net/rskec/credential-manifestapplicationv1.jsonld',
                            ],
                            type: ['VerifiablePresentation', 'CredentialApplication'],
                            credential_application: {
                                id: (0, utils_1.createUUID)(),
                                manifest_id: manifest_id,
                                format: {
                                    ldp_vc: {
                                        proof_type: proofTypes,
                                    },
                                },
                            },
                            presentation_submission: {
                                id: (0, utils_1.createUUID)(),
                                definition_id: definition_id,
                                descriptor_map: descriptor_map,
                            },
                            verifiableCredential: verifiableCredential,
                        };
                        return [4 /*yield*/, callbacks_1.callbacks[types_1.Actor.Holder].signPresentation({
                                contentToSign: messageData,
                                challenge: challenge,
                                message: message
                            })];
                    case 1:
                        signedData = _f.sent();
                        return [2 /*return*/, {
                                id: (0, utils_1.createUUID)(),
                                media_type: 'application/json',
                                format: 'dif/credential-manifest/application@v1.0',
                                data: {
                                    json: signedData,
                                },
                            }];
                }
            });
        });
    };
    OfferCredentialHandler = __decorate([
        (0, register_handler_decorator_1.RegisterHandler)(types_1.Actor.Holder, types_1.WACIMessageType.OfferCredential)
    ], OfferCredentialHandler);
    return OfferCredentialHandler;
}());
exports.OfferCredentialHandler = OfferCredentialHandler;
