// bitcoin-ts/build/main/lib/auth/instruction-sets/common/common.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const transaction_1 = require("../../../transaction");
const arithmetic_1 = require("./arithmetic");
const bitwise_1 = require("./bitwise");
const combinators_1 = require("./combinators");
const crypto_1 = require("./crypto");
const errors_1 = require("./errors");
const flow_control_1 = require("./flow-control");
const nop_1 = require("./nop");
const opcodes_1 = require("./opcodes");
const push_1 = require("./push");
const splice_1 = require("./splice");
const stack_1 = require("./stack");
const time_1 = require("./time");
__export(require("./arithmetic"));
__export(require("./bitwise"));
__export(require("./combinators"));
__export(require("./crypto"));
__export(require("./descriptions"));
__export(require("./encoding"));
__export(require("./errors"));
__export(require("./flow-control"));
__export(require("./nop"));
__export(require("./opcodes"));
__export(require("./push"));
__export(require("./signing-serialization"));
__export(require("./splice"));
__export(require("./stack"));
__export(require("./time"));
__export(require("./types"));
var ConsensusCommon;
(function (ConsensusCommon) {
    /**
     * A.K.A. `MAX_SCRIPT_ELEMENT_SIZE`
     */
    ConsensusCommon[ConsensusCommon["maximumStackItemLength"] = 520] = "maximumStackItemLength";
    ConsensusCommon[ConsensusCommon["maximumScriptNumberLength"] = 4] = "maximumScriptNumberLength";
    /**
     * A.K.A. `MAX_OPS_PER_SCRIPT`
     */
    ConsensusCommon[ConsensusCommon["maximumOperationCount"] = 201] = "maximumOperationCount";
    /**
     * A.K.A. `MAX_SCRIPT_SIZE`
     */
    ConsensusCommon[ConsensusCommon["maximumBytecodeLength"] = 10000] = "maximumBytecodeLength";
    /**
     * A.K.A. `MAX_STACK_SIZE`
     */
    ConsensusCommon[ConsensusCommon["maximumStackDepth"] = 1000] = "maximumStackDepth";
})(ConsensusCommon = exports.ConsensusCommon || (exports.ConsensusCommon = {}));
exports.undefinedOperation = () => ({
    undefined: combinators_1.conditionallyEvaluate((state) => errors_1.applyError(errors_1.AuthenticationErrorCommon.unknownOpcode, state)),
});
exports.checkLimitsCommon = (operation) => (state) => {
    const nextState = operation(state);
    return nextState.stack.length + nextState.alternateStack.length >
        ConsensusCommon.maximumStackDepth
        ? errors_1.applyError(errors_1.AuthenticationErrorCommon.exceededMaximumStackDepth, nextState)
        : nextState.operationCount > ConsensusCommon.maximumOperationCount
            ? errors_1.applyError(errors_1.AuthenticationErrorCommon.exceededMaximumOperationCount, nextState)
            : nextState;
};
exports.commonOperations = ({ flags, ripemd160, secp256k1, sha1, sha256, }) => {
    const unconditionalOperations = Object.assign(Object.assign(Object.assign({}, nop_1.disabledOperations()), push_1.pushOperations(flags)), combinators_1.mapOverOperations(flow_control_1.unconditionalFlowControlOperations(flags), combinators_1.incrementOperationCount));
    const conditionalOperations = combinators_1.mapOverOperations(Object.assign(Object.assign({}, push_1.pushNumberOperations()), { [opcodes_1.OpcodesCommon.OP_RESERVED]: flow_control_1.reservedOperation() }), combinators_1.conditionallyEvaluate);
    const incrementingOperations = combinators_1.mapOverOperations(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, arithmetic_1.arithmeticOperations(flags)), bitwise_1.bitwiseOperations()), crypto_1.cryptoOperations({
        flags,
        ripemd160,
        secp256k1,
        sha1,
        sha256,
    })), flow_control_1.conditionalFlowControlOperations()), stack_1.stackOperations(flags)), splice_1.spliceOperations()), time_1.timeOperations(flags)), nop_1.nonOperations(flags)), combinators_1.conditionallyEvaluate, combinators_1.incrementOperationCount);
    return combinators_1.mapOverOperations(Object.assign(Object.assign(Object.assign({}, unconditionalOperations), incrementingOperations), conditionalOperations), exports.checkLimitsCommon);
};
exports.cloneStack = (stack) => stack.reduce((newStack, element) => {
    // eslint-disable-next-line functional/no-expression-statement, functional/immutable-data
    newStack.push(element.slice());
    return newStack;
}, []);
exports.createAuthenticationProgramInternalStateCommon = (instructions, stack = []) => ({
    alternateStack: [],
    executionStack: [],
    instructions,
    ip: 0,
    lastCodeSeparator: -1,
    operationCount: 0,
    signatureOperationsCount: 0,
    stack,
});
exports.createAuthenticationProgramExternalStateCommon = (program) => ({
    correspondingOutput: program.inputIndex < program.spendingTransaction.outputs.length
        ? transaction_1.serializeOutput(program.spendingTransaction.outputs[program.inputIndex])
        : undefined,
    locktime: program.spendingTransaction.locktime,
    outpointIndex: program.spendingTransaction.inputs[program.inputIndex].outpointIndex,
    outpointTransactionHash: program.spendingTransaction.inputs[program.inputIndex]
        .outpointTransactionHash,
    outputValue: program.sourceOutput.satoshis,
    sequenceNumber: program.spendingTransaction.inputs[program.inputIndex].sequenceNumber,
    transactionOutpoints: transaction_1.serializeOutpoints(program.spendingTransaction.inputs),
    transactionOutputs: transaction_1.serializeOutputsForSigning(program.spendingTransaction.outputs),
    transactionSequenceNumbers: transaction_1.serializeSequenceNumbers(program.spendingTransaction.inputs),
    version: program.spendingTransaction.version,
});
exports.createAuthenticationProgramStateCommon = (instructions, stack, externalState) => (Object.assign(Object.assign({}, exports.createAuthenticationProgramInternalStateCommon(instructions, stack)), externalState));
exports.cloneAuthenticationProgramStateCommon = (state) => (Object.assign(Object.assign({}, (state.error === undefined ? {} : { error: state.error })), { alternateStack: state.alternateStack.slice(), correspondingOutput: state.correspondingOutput, executionStack: state.executionStack.slice(), instructions: state.instructions.slice(), ip: state.ip, lastCodeSeparator: state.lastCodeSeparator, locktime: state.locktime, operationCount: state.operationCount, outpointIndex: state.outpointIndex, outpointTransactionHash: state.outpointTransactionHash.slice(), outputValue: state.outputValue, sequenceNumber: state.sequenceNumber, signatureOperationsCount: state.signatureOperationsCount, stack: state.stack.slice(), transactionOutpoints: state.transactionOutpoints, transactionOutputs: state.transactionOutputs, transactionSequenceNumbers: state.transactionSequenceNumbers, version: state.version }));
/**
 * This is a meaningless but complete `CommonExternalProgramState`, useful for
 * testing and debugging.
 */
exports.createAuthenticationProgramExternalStateCommonEmpty = () => ({
    correspondingOutput: Uint8Array.of(1 /* correspondingOutput */),
    locktime: 0,
    outpointIndex: 0,
    outpointTransactionHash: new Uint8Array(32 /* length */).fill(5 /* outpointTransactionHash */),
    outputValue: 0,
    sequenceNumber: 0,
    transactionOutpoints: Uint8Array.of(2 /* transactionOutpoints */),
    transactionOutputs: Uint8Array.of(3 /* transactionOutputs */),
    transactionSequenceNumbers: Uint8Array.of(4 /* transactionSequenceNumbers */),
    version: 0,
});
/**
 * Create an "empty" CommonProgramState, suitable for testing a VM/compiler.
 */
exports.createAuthenticationProgramStateCommonEmpty = (instructions, stack = []) => (Object.assign(Object.assign({}, exports.createAuthenticationProgramInternalStateCommon(instructions, stack)), exports.createAuthenticationProgramExternalStateCommonEmpty()));
