// @extrimian/modena-sdk/dist/src/index.js
"use strict";
// NOTE: Aliases to classes and interfaces are used for external consumption.
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateCommitmentUtils = exports.LocalSigner = exports.ModenaRequest = exports.ModenaSdkConfig = exports.ModenaPublicKeyPurpose = exports.ModenaNetwork = exports.ModenaKey = exports.ModenaDid = void 0;
const ModenaDid_1 = require("./ModenaDid");
exports.ModenaDid = ModenaDid_1.default;
const ModenaKey_1 = require("./ModenaKey");
exports.ModenaKey = ModenaKey_1.default;
const ModenaNetwork_1 = require("./enums/ModenaNetwork");
exports.ModenaNetwork = ModenaNetwork_1.default;
const ModenaPublicKeyPurpose_1 = require("./enums/ModenaPublicKeyPurpose");
exports.ModenaPublicKeyPurpose = ModenaPublicKeyPurpose_1.default;
const ModenaRequest_1 = require("./ModenaRequest");
exports.ModenaRequest = ModenaRequest_1.default;
const ModenaSdkConfig_1 = require("./ModenaSdkConfig");
exports.ModenaSdkConfig = ModenaSdkConfig_1.default;
const LocalSigner_1 = require("./LocalSigner");
exports.LocalSigner = LocalSigner_1.default;
const UpdateCommitmentUtils_1 = require("./utils/UpdateCommitmentUtils");
Object.defineProperty(exports, "UpdateCommitmentUtils", { enumerable: true, get: function () { return UpdateCommitmentUtils_1.UpdateCommitmentUtils; } });
