// @decentralized-identity/ion-sdk/dist/lib/IonKey.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const InputValidator_1 = require("./InputValidator");
const randomBytes = require('randombytes');
const secp256k1 = require('@transmute/did-key-secp256k1');
/**
 * Class containing operations related to keys used in ION.
 */
class IonKey {
    /**
     * Generates SECP256K1 key pair to be used in an operation.
     * Mainly used for testing.
     * @returns [publicKey, privateKey]
     */
    static generateEs256kDidDocumentKeyPair(input) {
        return __awaiter(this, void 0, void 0, function* () {
            const id = input.id;
            const purposes = input.purposes;
            InputValidator_1.default.validateId(id);
            InputValidator_1.default.validatePublicKeyPurposes(purposes);
            const [publicKey, privateKey] = yield IonKey.generateEs256kKeyPair();
            const publicKeyModel = {
                id,
                type: 'EcdsaSecp256k1VerificationKey2019',
                publicKeyJwk: publicKey
            };
            // Only add the `purposes` property If given `purposes` array has at least an entry.
            if (purposes !== undefined && purposes.length > 0) {
                publicKeyModel.purposes = purposes;
            }
            return [publicKeyModel, privateKey];
        });
    }
    /**
     * Generates SECP256K1 key pair for ION operation use.
     * @returns [publicKey, privateKey]
     */
    static generateEs256kOperationKeyPair() {
        return __awaiter(this, void 0, void 0, function* () {
            const keyPair = yield IonKey.generateEs256kKeyPair();
            return keyPair;
        });
    }
    static generateEs256kKeyPair() {
        return __awaiter(this, void 0, void 0, function* () {
            const keyPair = yield secp256k1.Secp256k1KeyPair.generate({
                secureRandom: () => randomBytes(32)
            });
            const { publicKeyJwk, privateKeyJwk } = yield keyPair.toJsonWebKeyPair(true);
            return [publicKeyJwk, privateKeyJwk];
        });
    }
}
exports.default = IonKey;
