// @extrimian/kms-suite-bbsbls2020/src/services/bbsbls2020-suite.ts
import { IVCJsonLDKeyPair, Suite, suiteDecorator } from "@extrimian/kms-core";
import { JsonLDSuite } from "@extrimian/kms-suite-jsonld";
import {
  BbsBlsSignature2020,
  Bls12381G2KeyPair,
} from "@mattrglobal/jsonld-signatures-bbs";

@suiteDecorator(Suite.Bbsbls2020)
export class BbsBls2020Suite extends JsonLDSuite {
  async create(): Promise<IVCJsonLDKeyPair> {
    var keyPair = await Bls12381G2KeyPair.generate();

    this.secret = keyPair;

    return {
      privateKey: keyPair.privateKey,
      publicKey: keyPair.publicKey,
      keyType: keyPair.type
    };
  }

  protected async getSuite(params?: {
    verificationMethodId: string;
    controllerDid: string;
  }): Promise<any> {
    if (this.secret && params) {
      return new BbsBlsSignature2020({
        key: new Bls12381G2KeyPair({
          id: params.verificationMethodId,
          controller: params.controllerDid,
          publicKeyBase58: this.secret.publicKey,
          privateKeyBase58: this.secret.privateKey,
        }),
      });
    } else {
      return new BbsBlsSignature2020();
    }
  }
}
