// bitcoin-ts/build/main/lib/auth/templates/language/compile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("./errors");
const parse_1 = require("./parse");
const reduce_1 = require("./reduce");
const resolve_1 = require("./resolve");
var Formatting;
(function (Formatting) {
    Formatting[Formatting["requiresCommas"] = 3] = "requiresCommas";
    Formatting[Formatting["requiresOr"] = 2] = "requiresOr";
})(Formatting || (Formatting = {}));
/**
 * The constant used by the parser to denote the end of the input
 */
const EOF = 'EOF';
/**
 * A text-formatting method to pretty-print the list of expected inputs
 * (`Encountered unexpected input while parsing script. Expected ...`). If
 * present, the `EOF` expectation is always moved to the end of the list.
 * @param expectedArray - the alphabetized list of expected inputs produced by
 * `parseScript`
 */
const describeExpectedInput = (expectedArray) => {
    const newArray = expectedArray.filter((value) => value !== EOF);
    // eslint-disable-next-line functional/no-conditional-statement
    if (newArray.length !== expectedArray.length) {
        // eslint-disable-next-line functional/no-expression-statement, functional/immutable-data
        newArray.push('the end of the script');
    }
    const withoutLastElement = newArray.slice(0, newArray.length - 1);
    const lastElement = newArray[newArray.length - 1];
    return `Encountered unexpected input while parsing script. Expected ${newArray.length >= Formatting.requiresCommas
        ? withoutLastElement.join(', ').concat(`, or ${lastElement}`)
        : newArray.length === Formatting.requiresOr
            ? newArray.join(' or ')
            : lastElement}.`;
};
/**
 * Note, `compileScript` is the recommended API for using this method.
 */
exports.compileScriptText = ({ data, environment, script, scriptId, }) => {
    const parseResult = parse_1.parseScript(script);
    if (!parseResult.status) {
        return {
            errorType: 'parse',
            errors: [
                {
                    error: describeExpectedInput(parseResult.expected),
                    range: {
                        endColumn: parseResult.index.column,
                        endLineNumber: parseResult.index.line,
                        startColumn: parseResult.index.column,
                        startLineNumber: parseResult.index.line,
                    },
                },
            ],
            success: false,
        };
    }
    const resolver = resolve_1.createIdentifierResolver(scriptId, data, environment);
    const resolvedScript = resolve_1.resolveScriptSegment(parseResult.value, resolver);
    const resolutionErrors = errors_1.getResolutionErrors(resolvedScript);
    if (resolutionErrors.length !== 0) {
        return {
            errorType: 'resolve',
            errors: resolutionErrors,
            parse: parseResult.value,
            resolve: resolvedScript,
            success: false,
        };
    }
    const reduction = reduce_1.reduceScript(resolvedScript, environment.vm, environment.createState);
    return Object.assign(Object.assign({}, (reduction.errors === undefined
        ? { bytecode: reduction.bytecode, success: true }
        : { errorType: 'reduce', errors: reduction.errors, success: false })), { parse: parseResult.value, reduce: reduction, resolve: resolvedScript });
};
/**
 * Parse, resolve, and reduce the provided BTL script using the provided `data`
 * and `environment`.
 */
exports.compileScript = (scriptId, data, environment) => {
    const script = environment.scripts[scriptId];
    if (script === undefined) {
        return {
            errorType: 'parse',
            errors: [
                {
                    error: `No script with an ID of '${scriptId}' was provided in the compilation environment.`,
                    range: {
                        endColumn: 0,
                        endLineNumber: 0,
                        startColumn: 0,
                        startLineNumber: 0,
                    },
                },
            ],
            success: false,
        };
    }
    return exports.compileScriptText({
        data,
        environment,
        script,
        scriptId,
    });
};
