// bitcoin-ts/build/main/lib/auth/templates/language/reduce.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const format_1 = require("../../../format/format");
const opcodes_1 = require("../../instruction-sets/common/opcodes");
const instruction_sets_1 = require("../../instruction-sets/instruction-sets");
const instruction_sets_utils_1 = require("../../instruction-sets/instruction-sets-utils");
const pluckStartPosition = (range) => ({
    startColumn: range.startColumn,
    startLineNumber: range.startLineNumber,
});
const pluckEndPosition = (range) => ({
    endColumn: range.endColumn,
    endLineNumber: range.endLineNumber,
});
const mergeRanges = (ranges) => {
    const unsortedMerged = ranges.reduce(
    // eslint-disable-next-line complexity
    (merged, range) => (Object.assign(Object.assign({}, (range.startLineNumber < merged.startLineNumber
        ? pluckStartPosition(range)
        : range.startLineNumber === merged.startLineNumber &&
            range.startColumn < merged.startColumn
            ? pluckStartPosition(range)
            : pluckStartPosition(merged))), (range.endLineNumber > merged.endLineNumber
        ? pluckEndPosition(range)
        : range.endLineNumber === merged.endLineNumber &&
            range.endColumn > merged.endColumn
            ? pluckEndPosition(range)
            : pluckEndPosition(merged)))), ranges[0]);
    return Object.assign(Object.assign({}, pluckStartPosition(unsortedMerged)), pluckEndPosition(unsortedMerged));
};
const emptyReductionTraceNode = (range) => ({
    bytecode: Uint8Array.of(),
    range,
});
/**
 * Aggregate instructions to build groups of non-malformed instructions.
 */
// eslint-disable-next-line complexity
const aggregatedParseReductionTraceNodes = (nodes) => {
    const aggregations = [];
    // eslint-disable-next-line functional/no-let
    let ip = 0;
    // eslint-disable-next-line functional/no-let, init-declarations
    let incomplete;
    // eslint-disable-next-line functional/no-loop-statement
    for (const node of nodes) {
        const bytecode = incomplete === undefined
            ? node.bytecode
            : format_1.flattenBinArray([incomplete.bytecode, node.bytecode]);
        const range = incomplete === undefined
            ? node.range
            : mergeRanges([incomplete.range, node.range]);
        // eslint-disable-next-line functional/no-expression-statement
        incomplete = undefined;
        const parsed = instruction_sets_utils_1.parseBytecode(bytecode);
        // eslint-disable-next-line functional/no-conditional-statement
        if (parsed.length === 0) {
            // eslint-disable-next-line functional/no-expression-statement, functional/immutable-data
            aggregations.push({
                instructions: [],
                lastIp: ip,
                range,
            });
            // eslint-disable-next-line functional/no-conditional-statement
        }
        else if (instruction_sets_utils_1.authenticationInstructionsAreNotMalformed(parsed)) {
            // eslint-disable-next-line functional/no-expression-statement
            ip += parsed.length;
            // eslint-disable-next-line functional/no-expression-statement, functional/immutable-data
            aggregations.push({
                instructions: parsed,
                lastIp: ip,
                range,
            });
            // eslint-disable-next-line functional/no-conditional-statement
        }
        else {
            // eslint-disable-next-line functional/no-expression-statement
            incomplete = { bytecode, range };
        }
    }
    return Object.assign({ aggregations, success: true }, (incomplete === undefined
        ? undefined
        : {
            remainingBytecode: incomplete.bytecode,
            remainingRange: incomplete.range,
            success: false,
        }));
};
/**
 * Evaluate an array of `InstructionAggregation`s with the provided
 * `AuthenticationVirtualMachine`, matching the results back to their source
 * ranges.
 */
exports.evaluateInstructionAggregations = (aggregations, 
// eslint-disable-next-line @typescript-eslint/no-explicit-any
vm, getState) => {
    var _a, _b;
    const nonEmptyAggregations = aggregations.filter((aggregation) => aggregation.instructions.length > 0);
    const evaluationPlan = nonEmptyAggregations.reduce((plan, aggregation) => {
        const instructions = [...plan.instructions, ...aggregation.instructions];
        return {
            breakpoints: [
                ...plan.breakpoints,
                { ip: aggregation.lastIp, range: aggregation.range },
            ],
            instructions,
        };
    }, { breakpoints: [], instructions: [] });
    const trace = vm.stateDebug(getState(evaluationPlan.instructions));
    const samples = evaluationPlan.breakpoints.map((breakpoint) => ({
        range: breakpoint.range,
        state: trace[breakpoint.ip - 1],
    }));
    const firstInvalidSample = samples.findIndex((sample) => sample.state === undefined);
    const errorSample = (_a = samples[firstInvalidSample - 1]) !== null && _a !== void 0 ? _a : samples[firstInvalidSample];
    return errorSample === undefined
        ? {
            samples: samples,
            success: true,
        }
        : {
            errors: [
                {
                    error: errorSample.state === undefined
                        ? `Failed to reduce evaluation: vm.debug produced no valid program states.`
                        : `Failed to reduce evaluation: ${(_b = errorSample.state.error) !== null && _b !== void 0 ? _b : 'unknown error'}`,
                    range: errorSample.range,
                },
            ],
            samples,
            success: false,
        };
};
/**
 * Incrementally evaluate an array of `ScriptReductionTraceNode`s, returning a
 * trace of the evaluation and the resulting top stack item (`evaluationResult`)
 * if successful.
 *
 * @param nodes - an array of reduced nodes
 * @param vm - the `AuthenticationVirtualMachine` to use in the evaluation
 * @param getState - a method which should generate a new ProgramState given an
 * array of `instructions`
 */
// eslint-disable-next-line complexity
exports.sampledEvaluateReductionTraceNodes = (nodes, 
// eslint-disable-next-line @typescript-eslint/no-explicit-any
vm, getState) => {
    const parsed = aggregatedParseReductionTraceNodes(nodes);
    const evaluated = exports.evaluateInstructionAggregations(parsed.aggregations, vm, getState);
    if (parsed.success && evaluated.success) {
        const samples = evaluated.samples.length > 0
            ? evaluated.samples
            : [{ range: parsed.aggregations[0].range, state: getState([]) }];
        const lastSample = samples[samples.length - 1];
        const lastStackItem = lastSample.state.stack[lastSample.state.stack.length - 1];
        const evaluationResult = lastStackItem === undefined ? Uint8Array.of() : lastStackItem.slice();
        return {
            bytecode: evaluationResult,
            samples,
            success: true,
        };
    }
    return {
        bytecode: Uint8Array.of(),
        errors: [
            ...(parsed.success
                ? []
                : [
                    {
                        error: `A sample is malformed and cannot be evaluated: ${instruction_sets_utils_1.disassembleBytecode(opcodes_1.OpcodesCommon, parsed.remainingBytecode)}`,
                        range: parsed.remainingRange,
                    },
                ]),
            ...(evaluated.success ? [] : evaluated.errors),
        ],
        samples: evaluated.samples,
        success: false,
    };
};
/**
 * This method will throw an error if provided a `compiledScript` with
 * compilation errors. To check for compilation errors, use `getCompileErrors`.
 * @param compiledScript - the `CompiledScript` to reduce
 * @param vm - the `AuthenticationVirtualMachine` to use for evaluations
 * @param createState - a method which returns the base `ProgramState` used when initializing evaluations
 */
exports.reduceScript = (compiledScript, 
// eslint-disable-next-line @typescript-eslint/no-explicit-any
vm, createState) => {
    const source = compiledScript.map((segment) => {
        switch (segment.type) {
            case 'bytecode':
                return { bytecode: segment.value, range: segment.range };
            case 'push': {
                if (segment.value.length === 0) {
                    return emptyReductionTraceNode(segment.range);
                }
                const push = exports.reduceScript(segment.value, vm, createState);
                const bytecode = instruction_sets_1.encodeDataPush(push.bytecode);
                return Object.assign(Object.assign({ bytecode }, (push.errors === undefined ? undefined : { errors: push.errors })), { range: segment.range, source: [push] });
            }
            case 'evaluation': {
                if (segment.value.length === 0) {
                    return emptyReductionTraceNode(segment.range);
                }
                if (typeof vm === 'undefined' || typeof createState === 'undefined') {
                    return Object.assign({ errors: [
                            {
                                error: 'Both a VM and a createState method are required to reduce evaluations.',
                                range: segment.range,
                            },
                        ] }, emptyReductionTraceNode(segment.range));
                }
                const reductionTrace = exports.reduceScript(segment.value, vm, createState);
                const evaluated = exports.sampledEvaluateReductionTraceNodes(reductionTrace.source, vm, createState);
                const errors = [
                    ...(reductionTrace.errors === undefined ? [] : reductionTrace.errors),
                    ...(evaluated.success ? [] : evaluated.errors),
                ];
                return Object.assign(Object.assign({}, (errors.length > 0
                    ? Object.assign({ errors }, emptyReductionTraceNode(segment.range)) : {
                    bytecode: evaluated.bytecode,
                    range: segment.range,
                })), { samples: evaluated.samples, source: [reductionTrace] });
            }
            case 'comment':
                return emptyReductionTraceNode(segment.range);
            case 'error':
                return Object.assign({ errors: [
                        {
                            error: `Tried to reduce a BTL script with resolution errors: ${segment.value}`,
                            range: segment.range,
                        },
                    ] }, emptyReductionTraceNode(segment.range));
            default:
                return new Error(
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                `"${segment.type}" is not a known segment type.`);
        }
    });
    const reduction = source.reduce((all, segment) => (Object.assign({ bytecode: [...all.bytecode, segment.bytecode], ranges: [...all.ranges, segment.range] }, (all.errors !== undefined || segment.errors !== undefined
        ? {
            errors: [
                ...(all.errors === undefined ? [] : all.errors),
                ...(segment.errors === undefined ? [] : segment.errors),
            ],
        }
        : undefined))), { bytecode: [], ranges: [] });
    return Object.assign(Object.assign({}, (reduction.errors === undefined
        ? undefined
        : { errors: reduction.errors })), { bytecode: format_1.flattenBinArray(reduction.bytecode), range: mergeRanges(reduction.ranges), source });
};
