// @extrimian/did-core/dist/models/verification-method.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerificationMethodHelper = exports.VerificationMethodTypes = void 0;
var VerificationMethodTypes;
(function (VerificationMethodTypes) {
    VerificationMethodTypes["Ed25519VerificationKey2018"] = "Ed25519VerificationKey2018";
    VerificationMethodTypes["EcdsaSecp256k1VerificationKey2019"] = "EcdsaSecp256k1VerificationKey2019";
    VerificationMethodTypes["X25519KeyAgreementKey2019"] = "X25519KeyAgreementKey2019";
    VerificationMethodTypes["JsonWebKey2020"] = "JsonWebKey2020";
    VerificationMethodTypes["GpgVerificationKey2020"] = "GpgVerificationKey2020";
    VerificationMethodTypes["Bls12381G1Key2020"] = "Bls12381G1Key2020";
    VerificationMethodTypes["RsaVerificationKey2018"] = "RsaVerificationKey2018";
})(VerificationMethodTypes = exports.VerificationMethodTypes || (exports.VerificationMethodTypes = {}));
var VerificationMethodHelper = /** @class */ (function () {
    function VerificationMethodHelper() {
    }
    VerificationMethodHelper.getVMValue = function (vm) {
        if (vm.publicKeyBase58 != null) {
            return vm.publicKeyBase58;
        }
        else if (vm.publicKeyHex != null) {
            return vm.publicKeyHex;
        }
        else if (vm.publicKeyGpg != null) {
            return vm.publicKeyGpg;
        }
    };
    return VerificationMethodHelper;
}());
exports.VerificationMethodHelper = VerificationMethodHelper;
