// @extrimian/dwn-client/dist/services/inbox-consumer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InboxConsumer = void 0;
const axios_1 = __importDefault(require("axios"));
const enums_1 = require("../types/enums");
const utils_1 = require("../utils");
class InboxConsumer {
    constructor(url, did) {
        this.inboxURL = url;
        this.did = did;
    }
    async getMessages(filters) {
        try {
            const request = {
                id: (0, utils_1.createUUID)(),
                target: this.did,
                messages: [
                    {
                        descriptor: Object.assign({ method: enums_1.ThreadMethod.Query }, filters),
                    },
                ],
            };
            const { data } = await axios_1.default.post(this.inboxURL, request);
            return this.mapResponse(data);
        }
        catch (error) {
            console.error(error);
            throw new Error(error.message);
        }
    }
    mapResponse(response) {
        return response.replies.flatMap((reply) => reply.entries.flatMap((thread) => thread.map((message) => (Object.assign(Object.assign({}, message), { data: (0, utils_1.decodeMessage)(message) })))));
    }
}
exports.InboxConsumer = InboxConsumer;
